agent/connect/ca/provider_aws.go

```diff
@@ -0,0 +1,656 @@
+	}
+	return *csrPEM, nil
+}
+
+func (a *AWSProvider) loadCACerts() error {
+	input := &acmpca.GetCertificateAuthorityCertificateInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+	}
+	output, err := a.client.GetCertificateAuthorityCertificate(input)
+	if err != nil {
+		return err
+	}
+
+	if a.isPrimary {
+		// Just use the cert as a root
+		a.rootPEM = *output.Certificate
+	} else {
+		a.intermediatePEM = *output.Certificate
+		// TODO(banks) support user-supplied CA being a Subordinate even in the
+		// primary DC. For now this assumes there is only one cert in the chain
+		if output.CertificateChain == nil {
+			return fmt.Errorf("Subordinate CA %s returned no chain", a.arn)
+		}
+		a.rootPEM = *output.CertificateChain
+	}
+	return nil
+}
+
+func (a *AWSProvider) signCSRRaw(csr *x509.CertificateRequest, templateARN string, ttl time.Duration) (string, error) {
+	// PEM encode the CSR
+	var pemBuf bytes.Buffer
+	if err := pem.Encode(&pemBuf, &pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csr.Raw}); err != nil {
+		return "", err
+	}
+
+	return a.signCSR(pemBuf.String(), templateARN, ttl)
+}
+
+func (a *AWSProvider) signCSR(csrPEM string, templateARN string, ttl time.Duration) (string, error) {
+	_, signAlg, err := keyTypeToAlgos(a.config.PrivateKeyType, a.config.PrivateKeyBits)
+	if err != nil {
+		return "", err
+	}
+
+	issueInput := acmpca.IssueCertificateInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+		Csr:                     []byte(csrPEM),
+		SigningAlgorithm:        aws.String(signAlg),
+		TemplateArn:             aws.String(templateARN),
+		Validity: &acmpca.Validity{
+			Value: aws.Int64(int64(ttl.Seconds() / 86400.0)),
+			Type:  aws.String(acmpca.ValidityPeriodTypeDays),
+		},
+	}
+
+	issueOutput, err := a.client.IssueCertificate(&issueInput)
+	if err != nil {
+		return "", fmt.Errorf("error issuing certificate from PCA: %s", err)
+	}
+
+	// wait for certificate to be created
+	certInput := acmpca.GetCertificateInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+		CertificateArn:          issueOutput.CertificateArn,
+	}
+	for {
+		certOutput, err := a.client.GetCertificate(&certInput)
+		if err != nil {
+			if err.(awserr.Error).Code() != acmpca.ErrCodeRequestInProgressException {
+				return "", fmt.Errorf("error retrieving certificate from PCA: %s", err)
+			}
+		}
+
+		if certOutput.Certificate != nil {
+			return *certOutput.Certificate, nil
+		}
+
+		a.logger.Printf("[DEBUG] connect.ca.aws: certificate %s is not ready"+
+			", waiting %s to check again", *issueOutput.CertificateArn,
+			a.config.PollInterval)
+
+		select {
+		case <-a.stopCh:
+			// Provider discarded
+			a.logger.Print("[WARN] connect.ca.aws: provider instance terminated"+
+				" while waiting for a new certificate %s to be ready.",
+				issueOutput.CertificateArn)
+			return "", fmt.Errorf("provider terminated")
+		case <-time.After(a.config.PollInterval):
+			// Continue looping...
+		}
```

Comment:
I assume this request doesn't time out to AWS in practice but this could hang indefinitely if the request status stalled on their end for some reason, right? Maybe we should add a hard timeout here to fail the request after enough waiting. Ditto for the above CA request loop.

Reply:
I couldnt find anything in the AWS docs for these endpoints about having a hard timeout, and even if they do we might still want a cutoff on our end in case the status of the request stalls for some reason.

Reply:
Yeah this is a great call. Our own clients probably don't expect to wait that long although net/rpc doesn't actually have timeouts right now.

I think 30 seconds to sign a cert is probably suficient. I'm also thinking of getting rid of the fixed poll interval and instead do something like:
 - wait 20ms before first attempt since I've never seen an immediate one work first time
 - wait 500ms before second
 - wait 1 second before third, (hopefully majority complete this fast - i've always seen it work after first 5 second wait so far)
 - wait 3 seconds before fourth
 - wait 5 seconds for the next 5 tries, total time after that is ~30 seconds so just give up and return a timeout error. We can tweak these later, but having them user-configurable doesn't make much sense to me for now. Maybe the overall timeout might be configurable later.